Zombie/src/main/java/dk/gruppeseks/bodtrd/zombie/ZombieProcessor.java

```diff
@@ -21,35 +25,101 @@
+    private final int TIME_BETWEEN_PATH_UPDATE = 1500;
+    private final Map<Integer, Stack<Position>> zombiePath = new HashMap();
+    private final Map<Integer, Position> nextPos = new HashMap();
+    private final Map<Integer, Long> zombieLastPathUpdate = new HashMap();
+    private final Map<Integer, Position> zombieLastPosition = new HashMap();
+    private boolean checked = false;
+    private int zombieId;
 
-    public ZombieProcessor(Map<Integer, Entity> zombies)
+    public ZombieProcessor(Map<Integer, Entity> zombies, AISPI ai)
     {
         _zombies = zombies;
+        _ai = ai;
+
     }
 
     @Override
     public void process(World world)
     {
         Position playerPos = world.getGameData().getPlayerPosition();
+        Body playerBod = world.getGameData().getPlayerBody();
+        Position playerCenter = new Position(playerPos.getX() + playerBod.getWidth() / 2, playerPos.getY() + playerBod.getHeight() / 2);
+
         for (Entity zombie : _zombies.values())
         {
             Position zombiePos = zombie.get(Position.class);
+            Body zombieBod = zombie.get(Body.class);
+            Position zombieCenter = new Position(zombiePos.getX() + zombieBod.getWidth() / 2, zombiePos.getY() + zombieBod.getHeight() / 2);
             Velocity zombieVel = zombie.get(Velocity.class);
-            double dx = playerPos.getX() - zombiePos.getX();
-            double dy = playerPos.getY() - zombiePos.getY();
-            Vector2 velocity = new Vector2(dx, dy);
-            if (velocity.getMagnitude() < AGGRO_RANGE)
+            Vector2 velocity = new Vector2(zombieCenter, playerCenter);
+            if (playerCenter != null && velocity.getMagnitude() < DUMB_AI_RANGE)
             {
                 velocity.setMagnitude(MOVEMENT_SPEED);
             }
+            else if (playerCenter != null && velocity.getMagnitude() < AGGRO_RANGE)
+            {
+                Long lastUpdate = zombieLastPathUpdate.get(zombie.getID());
+                if (lastUpdate == null || (System.currentTimeMillis() - lastUpdate) > TIME_BETWEEN_PATH_UPDATE)
+                {
+                    try
+                    {
+                        zombieLastPathUpdate.put(zombie.getID(), System.currentTimeMillis());
+
+                        Stack<Position> path = _ai.getPath(zombieCenter, playerCenter, world);
+                        path.pop(); // We dont need the first route.
+                        zombiePath.put(zombie.getID(), path);
+                        velocity = new Vector2(zombieCenter, path.peek());
+                    }
+                    catch (NoPathException ex)
+                    {
+                        System.out.println("NO PATH");
+                        velocity.setMagnitude(MOVEMENT_SPEED);
+                    }
+                }
+                Stack<Position> path = zombiePath.get(zombie.getID());
+                if (zombieLastPosition.get(zombie.getID()) != null && path != null)
```

Comment:
whats the idea behind zombieLastPosition? by the looks of it it can be completely avoided